### ISSUE TITLE
Similar project

### ISSUE THREAD

**Issue by bfredl** [Sean1708/NeoVim.jl#1]
Nice work. Just to let you know, I've also been working on a neovim client [here](https://github.com/bfredl/Neovim.jl), though it is much work in progress. Essentially it currently supports the following:

```
using Neovim
import Neovim: get_buffers, set_line, get_line
nvim = nvim_spawn()
buf = get_buffers(nvim)[1]
@assert isa(buf,Buffer)
set_line(buf, 1, "some text")
text = get_line(buf, 1)
@assert text == "some text"
```

**Comment by Sean1708** [comment]
Hmmm, interesting. You seem to have gone with a slightly more C-style way of doing things but it looks quite good.

**Comment by bfredl** [comment]
Yes, what I've done so far is auto-generating a "C-style" interface from the api metadata. Then, later on, one could add high-level "julian" wrappers for instance

```
Base.setindex!(b::Buffer, line, i::Integer) = set_line(Buffer, i, line)
```

Your repo contains many interesting ideas on this, for instance,

```
Base.setindex!(b::Buffer, val, var::Symbol) = :buffer_set_var
Base.setindex!(b::Buffer, val, opt::ByteString) = :buffer_set_option
```

But I actually thought of doing it the other way around (for me `Symbol` represents something more builtin/static than a `String`)

**Comment by Sean1708** [comment]
Ah I see, you'd do a similar kind of thing to me.

Those were just some rough ideas I was mapping out, I probably won't end up having so many `set index!`s since it'll get confusing.

**Comment by Sean1708** [comment]
I just had another quick look at your project and noticed you were looking into write a NeoVim REPL. Funny thing is, I started this entire project because I wanted to write a REPL in NeoVim.

**Comment by bfredl** [comment]
Yes, I am imagining two different broad appoaches:
- Embed a REPL inside `nvim`. this is the approach taken by [ivanov/vim-ipython](https://github.com/ivanov/vim-ipython) and my nvim fork [nvim-ipy](https://github.com/bfredl/nvim-ipy). ( In principle IJulia should work just as well with these, but there is some crashes that I get with IJuila that are hard to debug due to the intermediate python layer, so I'm looking into a pure nvim+julia solution)
- Embed nvim inside a REPL. This I what I have prototyped in `Neovim.jl`. It actually works relatively well for just normal mode, but more work using [abstract_ui](https://github.com/neovim/neovim/pull/1657) will be neccesary to support visual mode, command line etc.

**Comment by Sean1708** [comment]
Just a quick point looking at your code, everything that needs to be compiled at runtime (i.e. everything in your `api_gen.jl`) needs to be put inside (or called from) `__init__()` otherwise you're going to run into problems when [module caching](https://github.com/JuliaLang/julia/pull/8745) comes round.

**Comment by Sean1708** [comment]
And `__init__()` is called automatically when the module loads.

**Comment by bfredl** [comment]
Either that, or somehow force regeneration of the cached module when api metadata changes (which should not be far too often, when `nvim` gets closer to stable). Right now, at least a warning will be printed in this situation.

**Comment by bfredl** [comment]
Also, using stagedfunctions, I hope to eliminate much of the code generation that happens on module load time.

**Comment by Sean1708** [comment]
Hmmm, how are you planning on using staged functions? If you're planning in rewriting `build_function` as a staged function then how are you planning to dispatch on types? If you're planning to build staged functions (i.e. the api would be made up of staged functions) then I don't think that'll be any simpler.

Either way the project seems to be coming along really well,  keep up the good work.

**Comment by bfredl** [comment]
Not sure, but one approach would be this:

```
stagedfunction api_call{method}(::Type{Val{method}}, recv::NvimObject, args...)
    fullname = string(prefix(recv), method)
    # check types of args with metadata for fullname
    ...
end
prefix(::Type{NvimClient}) = "vim_"
prefix(::Type{Buffer}) = "buffer_"
```

then a polymorphic function like `get_option` would be

```
get_option(recv::NvimObject, name) = api_call(Val{:get_option}, recv, name)
```
